File: chat-api/src/main/java/gg/warcraft/chat/api/profile/service/ChatProfileRepository.java
package gg.warcraft.chat.api.profile.service;

import gg.warcraft.chat.api.profile.ChatProfile;

import java.util.UUID;

/**
 * This repository is injectable.
 * <p>
 * The chat profile repository serves as a point of entry into the chat module implementation. Generally you want to go
 * through the chat profile query and command services to get and update chat profiles, but if the need arises for you
 * to forgo those services you can interact directly with the repository via this interface.
 */
public interface ChatProfileRepository {

    /**
     * Returns the chat profile of the specified player.
     *
     * @param playerId The id of the player.
     * @return The chat profile of the specified player. Can be null.
     */
    ChatProfile get(UUID playerId);

    /**
     * Saves the specified chat profile to this repository.
     * <p>
     * NOTE: This repository does not do any safety checks. By explicitly foregoing the ChatProfileCommandService and
     * using this interface directly it is up to you to make sure that there are no null or empty string values set on
     * the profile.
     *
     * @param profile The chat profile to save.
     */
    void save(ChatProfile profile);
}

File: chat/src/main/scala/gg/warcraft/chat/profile/ChatProfileRepository.scala
package gg.warcraft.chat.profile

import java.util.UUID

import com.typesafe.config.Config
import gg.warcraft.monolith.api.core.Repository

import scala.collection.mutable
import scala.concurrent.Future

object ChatProfileRepository {
  private val _profiles = mutable.Map[UUID, ChatProfile]()

  private var _defaultTag: String = _
}

class ChatProfileRepository(
    override protected implicit val dbConfig: Config
) extends Repository {
  import ChatProfileRepository._
  import db._

  def profiles: Map[UUID, ChatProfile] =
    _profiles.asInstanceOf[Map[UUID, ChatProfile]]

  def defaultTag: String = _defaultTag

  def load(playerId: UUID): Option[ChatProfile] =
    db.run(query[ChatProfile].filter(_.playerId == lift(playerId))).headOption

  def save(profile: ChatProfile): Option[ChatProfile] = {
    Future { db.run(query[ChatProfile].insert(lift(profile))) }
    _profiles.put(profile.playerId.get, profile)
  }
}
